google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,353 @@
+    // to ensure that the delete catchingAsync() callback is run before the flushFuture callback.
+    List<ApiFuture<Void>> pendingFutures = new ArrayList<>();
+
+    // Delete the provided document reference if one was provided.
+    if (documentReference != null) {
+      ApiFuture<Void> voidDeleteFuture =
+          ApiFutures.transformAsync(
+              catchingDelete(documentReference),
+              new ApiAsyncFunction<WriteResult, Void>() {
+                public ApiFuture<Void> apply(WriteResult result) {
+                  return ApiFutures.immediateFuture(null);
+                }
+              },
+              MoreExecutors.directExecutor());
+      pendingFutures.add(voidDeleteFuture);
+    }
+
+    ApiFuture<Void> flushFuture = writer.flush();
+    pendingFutures.add(flushFuture);
+
+    // Completes the future returned by run() and sets the exception if an error occurred.
+    ApiFutures.transformAsync(
+        ApiFutures.allAsList(pendingFutures),
+        new ApiAsyncFunction<List<Void>, Void>() {
+          public ApiFuture<Void> apply(List<Void> unused) {
+            if (lastError == null) {
+              completionFuture.set(null);
+            } else {
+              String message =
+                  errorCount
+                      + (errorCount != 1 ? " deletes" : " delete")
+                      + " failed. "
+                      + lastError.getMessage();
+              if (lastError instanceof FirestoreException) {
+                lastError =
+                    new FirestoreException(message, ((FirestoreException) lastError).getStatus());
+              } else {
+                lastError = new Throwable(message, lastError);
+              }
+              completionFuture.setException(lastError);
+            }
+            return ApiFutures.immediateFuture(null);
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  /** Deletes the provided reference and increments the error count if the delete fails. */
+  private ApiFuture<WriteResult> catchingDelete(DocumentReference reference) {
+    ApiFuture<WriteResult> deleteFuture = writer.delete(reference);
+    return ApiFutures.catchingAsync(
+        deleteFuture,
+        Throwable.class,
+        new ApiAsyncFunction<Throwable, WriteResult>() {
+          public ApiFuture<WriteResult> apply(Throwable e) {
+            incrementErrorCount(e);
+            return ApiFutures.immediateFuture(null);
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  /** Deletes the provided reference and starts the next stream if conditions are met. */
+  private void deleteReference(final DocumentReference reference) {
+    pendingOperationsCount++;
+    ApiFutures.transformAsync(
+        catchingDelete(reference),
+        new ApiAsyncFunction<WriteResult, Object>() {
+
+          public ApiFuture<Object> apply(WriteResult result) {
+            pendingOperationsCount--;
```

Comment:
If BulkWriter sends two requests in parallel, this code might be invoked from two different Gax threads. I think this needs synchronization (same with the error handler).